common/src/main/java/com/revenuecat/purchases/common/caching/DeviceCache.kt

```diff
@@ -270,28 +284,78 @@ open class DeviceCache(
 
     @Synchronized
     fun setOfferingsCacheTimestampToNow() {
-        offeringsCachedObject.updateCacheTimestamp(Date())
+        offeringsCachedObject.updateCacheTimestamp(dateProvider.now)
     }
 
-    // endregion
-
     private fun clearOfferingsCache() {
         offeringsCachedObject.clearCache()
     }
 
+    // endregion
+
+    // region ProductEntitlementMappings
+
+    @Synchronized
+    fun cacheProductEntitlementMappings(productEntitlementMappings: ProductEntitlementMappings) {
+        preferences.edit()
+            .putString(
+                productEntitlementMappingCacheKey,
+                productEntitlementMappings.toJson().toString()
+            ).apply()
+
+        setProductEntitlementMappingCacheTimestampToNow()
+    }
+
+    @Synchronized
+    fun setProductEntitlementMappingCacheTimestampToNow() {
+        setProductEntitlementMappingCacheTimestamp(dateProvider.now)
+    }
+
+    private fun setProductEntitlementMappingCacheTimestamp(date: Date) {
+        preferences.edit().putLong(productEntitlementMappingLastUpdatedCacheKey, date.time).apply()
+    }
+
+    @Synchronized
+    fun isProductEntitlementMappingsCacheStale(): Boolean {
+        return getProductEntitlementMappingsLastUpdated().isStale(PRODUCT_ENTITLEMENT_MAPPING_CACHE_REFRESH_PERIOD)
+    }
+
+    @Synchronized
+    fun getProductEntitlementMappings(): ProductEntitlementMappings? {
+        return preferences.getString(productEntitlementMappingCacheKey, null)?.let { jsonString ->
+            return ProductEntitlementMappings.fromJson(JSONObject(jsonString))
+        }
+    }
+
+    private fun getProductEntitlementMappingsLastUpdated(): Date? {
+        return if (preferences.contains(productEntitlementMappingLastUpdatedCacheKey)) {
+            Date(preferences.getLong(productEntitlementMappingLastUpdatedCacheKey, -1))
+        } else {
+            null
+        }
+    }
+
+    // endregion
+
+    // region utils
+
     private fun Date?.isStale(appInBackground: Boolean): Boolean {
-        return this?.let { cachesLastUpdated ->
+        return this?.let {
             log(LogIntent.DEBUG, ReceiptStrings.CHECKING_IF_CACHE_STALE.format(appInBackground))
```

Comment:
I thought about also logging this for the new cache, but honestly I'm not sure how useful this is... Any thoughts about removing this log? For now I decided to leave it as is @vegaro